node/src/components/linear_chain.rs

```diff
@@ -197,28 +269,61 @@ where
-                let mut effects = effect_builder.put_execution_results_to_storage(block_hash, execution_results).ignore();
+                let mut effects = effect_builder
+                    .put_execution_results_to_storage(block_hash, execution_results)
+                    .ignore();
                 effects.extend(
-                    effect_builder.handle_linear_chain_block(block_header.clone())
-                    .event(move |signature| Event::NewFinalitySignature(block_hash, signature)));
-                effects.extend(effect_builder.announce_block_added(block_hash, block_header).ignore());
+                    effect_builder
+                        .handle_linear_chain_block(block_header.clone())
+                        .event(move |fs| Event::NewFinalitySignature(Box::new(fs))),
+                );
+                effects.extend(
+                    effect_builder
+                        .announce_block_added(block_hash, block_header)
+                        .ignore(),
+                );
                 effects
-            },
-            Event::NewFinalitySignature(block_hash, signature) => {
+            }
+            Event::NewFinalitySignature(fs) => {
+                let FinalitySignature {
+                    block_hash,
+                    signature,
+                    public_key,
+                } = *fs;
+                let verified =
+                    crypto::asymmetric_key::verify(block_hash.inner(), &signature, &public_key);
+                if let Err(err) = verified {
+                    error!(%block_hash, %public_key, %err, "received invalid finality signature");
+                    return Effects::new();
+                }
+                debug!(%block_hash, %public_key, "received new finality signature");
                 effect_builder
                     .get_block_from_storage(block_hash)
-                    .then(move |maybe_block| match maybe_block {
-                        Some(mut block) => {
-                            block.append_proof(signature);
-                            effect_builder.put_block_to_storage(Box::new(block))
-                        }
-                        None => {
-                            warn!("Received a signature for {} but block was not found in the Linear chain storage", block_hash);
-                            panic!("Unhandled")
-                        }
+                    .event(move |maybe_block| {
+                        Event::BlockSignatureContinuation(Box::new(maybe_block), fs)
                     })
-                    .ignore()
+            }
+            Event::BlockSignatureContinuation(maybe_block, fs) => match *maybe_block {
+                None => {
+                    let block_hash = fs.block_hash();
+                    let public_key = fs.public_key();
+                    warn!(%block_hash, %public_key, "received a signature for a block that was not found in storage");
```

Comment:
I wonder whether we should keep the signature in a cache for a while, or request the missing block from the sender, or somehow just re-send the signature after a while. Otherwise it could happen that gossiping a signature simply fails, and nobody ever collects a quorum of them.
(Not necessarily in this PR.)

Reply:
I agree but I hope/think this will be handled by the new networking's gossiping logic. Instead of calling `broadcast(message)` I will call `gossip(message)` and it will take care of all that.